synth/snsynth/preprocessors/data_transformer.py

```diff
@@ -0,0 +1,196 @@
+    Discrete columns are encoded using a scikit-learn OneHotEncoder.
+    """
+
+    def __init__(self, epsilon):
+        """Create a data transformer.
+
+        Args:
+            max_clusters (int):
+                Maximum number of Gaussian distributions in Bayesian GMM.
+            epsilon (float):
+                epsilon for DP StandardScaler preprocessor
+        """
+        self.epsilon = epsilon
+
+    def _fit_continuous(self, column_name, raw_column_data):
+        """Fit DP Standard Scaler for continuous column."""
+        scaler = StandardScaler(epsilon=self.epsilon)
+        scaler.fit(raw_column_data.reshape(-1, 1))
+
+        return ColumnTransformInfo(
+            column_name=column_name,
+            column_type="continuous",
+            transform=scaler,
+            transform_aux=None,
+            output_info=[SpanInfo(1, 'tanh')],
+            output_dimensions=1)
+
+    def _fit_discrete(self, column_name, raw_column_data):
+        """Fit one hot encoder for discrete column."""
+        ohe = OneHotEncodingTransformer()
+        ohe.fit(raw_column_data)
+        num_categories = len(ohe.dummies)
+
+        return ColumnTransformInfo(
+            column_name=column_name,
+            column_type="discrete",
+            transform=ohe,
+            transform_aux=None,
+            output_info=[SpanInfo(num_categories, 'softmax')],
+            output_dimensions=num_categories)
+
+    def fit(self, raw_data, discrete_columns=tuple()):
+        """Fit DP StandardScaler for continuous columns and One hot encoder for discrete columns.
+
+        This step also counts the #columns in matrix data, and span information.
+        """
+        self.output_info_list = []
+        self.output_dimensions = 0
+
+        if not isinstance(raw_data, pd.DataFrame):
+            self.dataframe = False
+            raw_data = pd.DataFrame(raw_data)
+        else:
+            self.dataframe = True
+
+        self._column_raw_dtypes = raw_data.infer_objects().dtypes
+
+        self._column_transform_info_list = []
+        for column_name in raw_data.columns:
+            raw_column_data = raw_data[column_name].values
+            if column_name in discrete_columns:
+                column_transform_info = self._fit_discrete(
+                    column_name, raw_column_data)
+            else:
+                column_transform_info = self._fit_continuous(
+                    column_name, raw_column_data)
+
+            self.output_info_list.append(column_transform_info.output_info)
+            self.output_dimensions += column_transform_info.output_dimensions
+            self._column_transform_info_list.append(column_transform_info)
+
+    def _transform_continuous(self, column_transform_info, raw_column_data):
+
+        scaler = column_transform_info.transform
+        normalized_values = scaler.transform(raw_column_data)
+        normalized_values = normalized_values/4
```

Comment:
Why are we dividing by 4 here and then undoing that later?

Reply:
original CTGAN paper was done similarly but with GMM. I also experimented with it, if we divide by 4 and then undoing it later, the distribution of fake data way more aligns with the original data distirbution

Reply:
Huh - that's odd!